src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs

```diff
@@ -451,12 +449,7 @@ public static RoutePatternPathSegment Segment(params RoutePatternPart[] parts)
                 throw new ArgumentNullException(nameof(parts));
             }
 
-            return SegmentCore(parts);
-        }
-
-        private static RoutePatternPathSegment SegmentCore(IEnumerable<RoutePatternPart> parts)
-        {
-            return new RoutePatternPathSegment(parts.ToArray());
+            return new RoutePatternPathSegment(parts);
```

Comment:
The reason `ToArray` is here is to make a copy of the array. Without the copy someone could pass an array to `Segment` and then modify it later. An extreme example would be an array from array pool that constantly changes. One of RoutePattern's goals is immutability.

I think at least there should be a comment added to make it clear to developers in the future that it is being copied on purpose.